File: solutions/sonnet-1/sonnet/ruby/anagrams.rb
words = STDIN.read.split("\n").map(&:strip).reject(&:empty?)

groups = Hash.new { |h, k| h[k] = [] }
words.each do |word|
  key = word.chars.sort.join
  groups[key] << word
end

lines = groups.values.map { |g| g.sort }
lines.sort_by! { |g| g.first }

puts lines.map { |g| g.join(" ") }

File: solutions/sonnet-1/sonnet/javascript/anagrams.js
const lines = require('fs').readFileSync(0, 'utf8').split('\n').filter(w => w.length > 0);

const groups = new Map();
for (const word of lines) {
  const key = word.split('').sort().join('');
  if (!groups.has(key)) groups.set(key, []);
  groups.get(key).push(word);
}

const result = [];
for (const words of groups.values()) {
  words.sort();
  result.push(words);
}

result.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));

console.log(result.map(g => g.join(' ')).join('\n'));
